### DISCUSSION TITLE
Passing arguments with lambda functions

### DISCUSSION THREAD

**Discussion by laserir** [zauberzeug/nicegui#471] [Q&A] [answered]
Hi all,

I create buttons dynamically after loading a json file like so:

```py
i = 0
for users in settings["users"]:
     user_buttons[i] = ui.button(users['name'], on_click=lambda: set_userto(users['id']))
     i = i+1
```

Problem is ALL buttons call set_userto(**4**)

Any ideas?

**Comment by rodja** [comment, marked as answer]
You need to capture the value of users at the time each button is created:

```py
for i, user in enumerate(settings["users"]):
     user_buttons[i] = ui.button(user['name'], on_click=lambda e, user=user: set_userto(user['id']))
```

**Reply by falkoschindler** [reply]
Note that this is independent of NiceGUI. Lambda statements don't capture the current value of variables in a for loop:

```py
callbacks = []
for letter in ['A', 'B', 'C']:
    callbacks.append(lambda: print(letter))

for callback in callbacks:
    callback() # output: C C C
```

Like @rodja says you need to capture the current `letter`:

```py
callbacks = []
for letter in ['A', 'B', 'C']:
    callbacks.append(lambda letter=letter: print(letter))

for callback in callbacks:
    callback() # output: A B C
```

**Comment by laserir** [comment]
Thanks! Anyhow... now it throws me this exception when the on_click is triggered (creation of the buttons still works fine):
`TypeError: 'ClickEventArguments' object is not subscriptable`

**Reply by falkoschindler** [reply]
Oh, there's a little mistake in @rodja's code: `lambda user=user:` overwrites the optional event arguments a button click handler expects. Replace it with something like `lambda e, user=user:` and it should work.

It's sometimes difficult to help with incomplete code examples because we can't easily try them out.

Edit: I fixed Rodja's code.

**Reply by laserir** [reply]
Thank you guys so much!!! Really appreciate your work!
